### DISCUSSION TITLE
Modern CSS and standards compliance

### DISCUSSION THREAD

**Discussion by thescientist13** [uswds/uswds#4312] [Ideas]
Although I'm not super familiar with the full extent of what SASS usages is like throughout the USWDS aside from just variables and nesting, but would the USWDS team open to modernizing some aspects of the CSS architecture?

With a lot of the recent advances in CSS, especially with the introduction of [Custom properties (variables)](https://developer.mozilla.org/en-US/docs/Web/CSS/--*), there are slowly become less and less reasons to need a CSS pre-processor.  And with the help of tools like PostCSS, backfilling SASS support for something like say [nested CSS](https://github.com/postcss/postcss-nested), is fairly straightforward these days for someone using a Node / npm workflow.

In general, like with the discussion happening around adopting a base library for components ( #2884), it seems like moving in a direction of standards compliance with CSS too would also be great both strategically and tactically for USWDS.  One great advantage is less tooling needed to author and maintain in the USWDS and also not having to deal with tools like SASS as they come and go.  In addition, browsers are getting even [closer now to being able to handle CSS using `import` / `export` (ESM)](https://twitter.com/justinfagnani/status/1403436100132110340) so even more DX that can just be straight offloaded to the browser!  🤩 

----

Certainly not under any illusions of the scope of an initiative like this as being trivial or anything, but no better time to start the conversation than the present.

Curious to hear what the team's thoughts are.  Thanks!  ✌️

**Comment by mejiaj** [comment]
Thanks for starting this discussion. It's something we've certainly considered. But we're not using SCSS just to get nesting. We use many functions and mixins to help us keep the styles maintainable. We're also very interested in SCSSmodules and how it can help build a leaner USWDS package.

You're right that an initiative like this is quite a lot of work. Something that we can't really dedicate the time to prototyping and testing. But definitely, something to think about for the future. Right now it seems like a lateral move.

The re-work of dropping SCSS in favor of PostCSS or plain CSS isn't justifiable right now. If you have any suggestions on how it can improve the codebase or dev experience; please share.

**Reply by thescientist13** [reply]
Thank you for sharing more from the project's perspective and apologies for the delayed follow up.

For a little context, I have been working on an [OSS project of my own](https://github.com/ProjectEvergreen/greenwood) and USWDS is, as a civic minded developer, one of my self identified "high impact" community component libraries that I want to [try and support](https://github.com/ProjectEvergreen/greenwood/discussions/523) in an effort to help advance web dev and web standards for everyone.  I have observed a potential need for continued reliance on SASS (Bootstrap v4) for one of my own web projects, so this feedback here is useful in that continuing to work on a [SASS plugin](https://github.com/AnalogStudiosRI/www.analogstudios.net/pull/15) could be useful at large.  😃 

Anyway, hope it is OK for me to share a bit of that context and have been keeping an eye on the 3.0 Beta as well and will definitely share and contribute back any feedback as I work through my examples of using USWDS!  👍

**Reply by mejiaj** [reply]
Totally! That insight is appreciated.

It's great that your project is supporting web components. We've prototyped with lit-html and have stenciljs next on our list. If you have any thoughts on either or web components  in general we'd love to hear them.

**Reply by thescientist13** [reply]
That's great, thanks!

Also would want to add that I am a member of an open source team called the [Web Components Community Group](https://web-components-cg.netlify.app/) that's made up of a cross section of WC and WC adjacent library authors and tooling creators throughout the JS ecosystem (like Lit and Stencil!)  

We just recently presented [a report at TPAC 2021](https://w3c.github.io/webcomponents-cg/) about some of the specs / features that we think would best help make an impact towards overall WC ergonomics and adoption.  If you have any questions or want to participate in similar standards advancement, there is [WCCG Slack](https://join.slack.com/t/webcomponentcommunity/shared_invite/zt-uinb2qud-cDvio5cC7IllxZfcTFf4QQ), as well as one for [Lit & Friends](https://join.slack.com/t/lit-and-friends/shared_invite/zt-llwznvsy-LZwT13R66gOgnrg12PUGqw), so feel free to join and participate!  We meet regularly and are in the process of thinking of how to redesign [webcomponents.org](https://www.webcomponents.org/) amongst other activities!

**Reply by thescientist13** [reply]
> The re-work of dropping SCSS in favor of PostCSS or plain CSS isn't justifiable right now. If you have any suggestions on how it can improve the codebase or dev experience; please share.

Not sure I actually answered this so just for posterity, the main reasons are technical flexibility and ease of developer support from USWDS as a community / maintainer.  As with most good things in technology, tools come and go, but standards pretty much stay around forever.  Like there was LESS before SASS, or CoffeeScript before ES6 before TypeScript, etc, so too I am sure there will be something after SASS.

And even SASS itself will have moving targets, like the migration to Dart SASS.  

Now, the main point here isn't to avoid all frameworks and tools, you support Node.js / npm workflows which is great, but also there is a strong governance behind both those projects and very clear lines of LTS.  In addition, they are foundational technologies for DX, as they just simply occupy the runtime and package manager space, and are themselves not packages / libraries.

So why does this matter to a _design system_?  IMO, the main reason is to help the _USWDS_ from having to chase a moving target every few years, or having to manage those upstream roadmaps and their impact on USWDS's own roadmap and maintenance efforts, which is to deliver a quality design system.  By adopting as much from the (web / NodeJS) platform as possible, USWDS can:
1. Remove the design system from the conversation of how does it work with "XYZ" framework.  Any good [modern JS framework](https://custom-elements-everywhere.com/) should handle HTML (e.g. Web Components) and CSS out of the box because well, our shared runtime is the browser.  This also allows teams to consume something like a theme into any custom processing system of their own. (I should add for CMS environments that are more constrained, just dropping in .css file without have to pre-process is also very very nice).
1.  Ease the burden of support for maintainers and onramp for beginners.  As per the above point, if USWDS is no longer opinionated in the tooling / technology it targets, a lot of the "tech support" can be offloaded to the MDN docs as effectively the build time and run time are both just common platforms that everyone has access to.  Not only does this help encourage first principals, the knowledge base built up around conversations like this can be relevant across technology stacks and can be preserver relevance much longer.

Also, I think it's just nice that if someone really wants to go vanilla to create a site, they should.  The web is so powerful these days, users deserve better than having to always be greeted with a frontend framework every modern website they visit.  I think web development has evolved enough now that there is a legitimate case to have completely framework free websites.  Not every use case, but a lot more than there were 5-10 years ago!  😃 🏆 

----

Naturally all features might not be there yet for the USWDS's use case specifically, but as per https://github.com/uswds/uswds/discussions/4443, perhaps just start with design tokens, then work your way up over a release or two?  Anything that can be shared at the common denominator I think is a net positive, IMO.

**Comment by jcklpe** [comment]
I like SCSS but I also think it would be useful to use things like CSS variables, which would also help move the USWDS towards being more tech stack agnostic. And there's no reason one can't use both SCSS and CSS variables etc.

**Reply by zarahzachz** [reply]
![both](https://user-images.githubusercontent.com/5159392/149829778-179bbfce-abb9-4e79-bc16-a964e229a479.gif)

I've seen several folks advocate for using both Sass and custom properties - mainly to get the best of both worlds. Custom properties do make theming and general variable manipulation much easier using Javascript, but Sass provides so many useful tools in addition to taking the compilation burden off of browsers. As Internet Person Andy Bell's said in the past:

> It’s probably better to cautiously pre-compile your CSS and not force the browser to work as hard. It’s already working hard enough, re-rendering the DOM every time one piece of data changes (cough reactive frameworks cough), so why make it work harder just so you can have native nesting? It smells like developer experience over user experience to me.

**Comment by ExMember** [comment]
I would be ecstatic if USWDS shifted to use custom properties (CSS variables) instead of SASS variables.

I have completely removed Node from multiple projects of mine. This brings me great joy because it makes the projects simpler, faster, and easier to maintain.

I still use USWDS, but rely on the compiled CSS/JS files. The downside is that I don't have a way to do any theming of USWDS. If the theming was done with CSS variables instead of in SASS I could still customize the theme without having to add a whole NPM pipeline to the project.

**Comment by thescientist13** [comment]
Coming out of the monthly call today, just wanted to note that [Constructable Stylesheets](https://developer.mozilla.org/en-US/docs/Web/API/CSSStyleSheet/CSSStyleSheet) are part of Baseline 2023, as an alternative to having all the CSS in JS files, and could be worth exploring from a perf / DX point of view.
```js
import styles from './uswds-banner.css' with { type: 'css' };
import theme from '../styles/theme.css' with { type: 'css' };

class Banner extends LitLElement {
  static styles = [ theme, styles ];

  // ...
}
```

Also nice that with this, you can author in _.css_ files to get that nice IDE / LSP integration for free.

**Comment by jcandan** [comment]
#4644
